[15.0][MIG] web_timeline: Migration to 15.0

cesar-tecnativa opened the pull request:
cc @Tecnativa TT35806

lmignon committed:
[IMP] web_timeline: Extracted from incubator

Commit:
[IMP] web_timeline: Several improvements:

* Avoid to display all items of group_by model
* Remove unnecessary readgroup
* Remove dependency on project. Modify module structure. Imporve readme file.
* Add setup.py

hbrunn committed:
[IMP] web_timeline: add few features

pedrobaeza committed:
[MIG] web_timeline: Migration to 9.0

ThomasBinsfeld committed:
[FIX] web_timeline: fix commas in manifest file

JBF91 committed:
[10.0] [FIX] [web_timeline] Correct write when grouped by (#678)

* [FIX] Correct write when grouped by

* Remove blank lines

* Minor changes

Creamaster committed:
[10.0][IMP] web_timeline improvements:

* Update the vis.js framework to latest stable version (v4.20.1) to add new functionality and save almost
400ko of source code in odoo assets (the vis dev team has isolate the code for timeline widget from the
 rest of their widgets)

* Add the possibility to display events with no duration (instantaneous) as the same time has regular events

* Add the possibility to choose between two scroll mode :
  -  zoom in time resolution (default)
  or
  -  vertical scroll (web browser default behaviour)

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[Fix OCA/project#317] Add the property date_delay to TimeLine views (same as base Calendar view)

In Odoo calendar view, the attribute date_delay is an alternative to date_stop, to provides the duration of the event instead of its end date.
This commit integrate this attribute to the Timeline view, but contrary to the base Calendar view the Timeline view could use both arguments (date_stop and date_delay) simultaneously.

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[IMP] Add an option to initialize the timeline window on display.

The default window display all the events (aka 'fit'), in case of events spread over the year, the events are invisibles (too small to be readable)

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[IMP] change the option to initialize the timeline window on display.

Change the attribute 'default_window' to 'mode' like the Odoo calendar view (to be iso functional)

npiganeau committed:
Fixed web_timeline.css

CSS classes have changed with the updated version of vis.js

Maartincm committed:
[MIG] web_timeline module from v10 to v11

* Update Version in Manifest
* Remove enconding in .py files
* Rewrite the view definition according to version 11

Maartincm committed:
[FIX] Missing GroupClick Functionality

tarteo committed:
[FIX] web_timeline: Make colors attribute optional (#969)

tarteo committed:
[IMP] web_timeline: Templates implementation like kanban

[IMP] web_timeline: Templates implementation like kanban

[ADD] Check if template exists

Update README.rst

tarteo committed:
[ADD] field_utils to web_timeline

[UPD] Version number

[UPD] README

tarteo committed:
[ADD] web_timeline: New dependency_arrow attribute

Update README.rst

[FIX] Remove console.log

[ADD] Make timeline.fit optional

[FIX] Use stringified points

[IMP] Reversed the arrow head and fixed lint issues

[IMP] Use options parameter for line color and width

[FIX] Version number

[IMP] Minor improvements

tarteo committed:
[ADD] Enable multiselect

[IMP] Prevent flickering

[IMP] Fixed headers

[ADD] Create new record by date range selection

Remove roadmap item from readme

[FIX] minimal height

tarteo committed:
[MIG] web_timeline: Migration to 12.0

tarteo committed:
[IMP] Readme: Available attribute list

Commit:
[IMP] Add stack support

Commit:
[FIX] web_timeline: fix readme - To be squashed with previous one on migration (#1137)

cvinh committed:
[web_timeline][IMP] Add margin support in web_timeline (#1135)

tarteo committed:
[FIX] web_timeline: Reloading view

[FIX] web_timeline: Reloading view

[FIX] web_timeline: Reloading view

denislour committed:
[12.0][FIX]web_timeline: fix group orders

denislour committed:
[12.0][web_timeline] Fix bug from group_order: locale instead locate.

hpatelserpentcs committed:
[FIX][12.0] web_timeline: Scroll Issue.

thoongnv committed:
[IMP] web_timeline: black, isort

thoongnv committed:
[MIG] web_timeline: Migration to 13.0

Tardo committed:
[MIG] web_timeline: Finish migration to 13.0

Tardo committed:
[FIX] web_timeline: Change vis-timeline library import

Tardo committed:
[IMP] web_timeline: CSS vis items

Tardo committed:
[FIX] web_timeline: Change draw_dependencies

Murtaza-OSI committed:
[FIX] Travis

Murtaza-OSI committed:
[14.0][MIG] Migrated web_timeline.

cesar-tecnativa committed:
[IMP] web_timeline: black, isort, prettier

cesar-tecnativa committed:
[MIG] web_timeline: Migration to 15.0

cesar-tecnativa commented:
@CarlosRoca13 please review

cesar-tecnativa commented:
@Tardo please review

pedrobaeza commented:
Have you tested it in a real case like with `project_timeline`?

cesar-tecnativa commented:
> Have you tested it in a real case like with `project_timeline`?

This is the goal https://github.com/OCA/project/pull/942

CarlosRoca13 commented on web_timeline/__manifest__.py:
```diff
@@ -0,0 +1,32 @@
+    "maintainers": ["tarteo"],
+    "application": False,
+    "installable": True,
+    "web.assets_backend": [
+        "web_timeline/static/src/scss/web_timeline.scss",
+        "web_timeline/static/src/js/timeline_view.js",
+        "web_timeline/static/src/js/timeline_renderer.js",
+        "web_timeline/static/src/js/timeline_controller.js",
+        "web_timeline/static/src/js/timeline_model.js",
+        "web_timeline/static/src/js/timeline_canvas.js",
+    ],
```
```suggestion
    "assets": {
        "web.assets_backend": [
            "web_timeline/static/src/scss/web_timeline.scss",
            "web_timeline/static/src/js/timeline_view.js",
            "web_timeline/static/src/js/timeline_renderer.js",
            "web_timeline/static/src/js/timeline_controller.js",
            "web_timeline/static/src/js/timeline_model.js",
            "web_timeline/static/src/js/timeline_canvas.js",
        ]
    },
```

cesar-tecnativa commented:
@sergio-teruel plesae review

hbrunn committed:
[FIX] web_timeline: Don't crash when restoring timeline from breadcrumb

cesar-tecnativa commented:
@LudLaf please review

jguenat commented on web_timeline/__manifest__.py:
```diff
@@ -0,0 +1,34 @@
+            "web_timeline/static/src/js/timeline_controller.js",
+            "web_timeline/static/src/js/timeline_model.js",
+            "web_timeline/static/src/js/timeline_canvas.js",
+        ]
```
```suggestion
        ],
        "web.assets_qweb": [
            "web_timeline/static/src/xml/web_timeline.xml",
        ],
```

jguenat commented on web_timeline/static/src/js/timeline_view.js:
```diff
@@ -0,0 +1,186 @@
+            Model: TimelineModel,
+            Controller: TimelineController,
+            Renderer: TimelineRenderer,
+        }),
```
```suggestion
        }),
        viewType: 'timeline',
```

jguenat reviewed (commented):
Hello,
I tested this with project_timeline.
As discussed there, Group by many2many is not supported by the timeline view.
Anyone know if it's easy to fix?

jguenat commented:
Also, what do you think about updating vis-timeline-graph2d lib to V 7.7?
https://unpkg.com/vis-timeline@7.7.0/peer/umd/vis-timeline-graph2d.min.js
https://unpkg.com/vis-timeline@7.7.0/styles/vis-timeline-graph2d.min.css

chrisandrewmann commented:
Issue with grouping by many2many fields causes crash as mentioned here.
https://github.com/OCA/project/pull/942#issuecomment-1198966193

CarlosRoca13 committed:
[FIX] web_timeline: Allow groupby m2m fields

chrisandrewmann commented:
@CarlosRoca13 Thanks for that, awesome job with adding the m2m field grouping. I had tried myself but javascript is not my strongest point.

CarlosRoca13 commented:
Hi! I added the changes added on https://github.com/OCA/web/pull/2256

I have also corrected the implementation to allow grouping by m2m fields with some restrictions:
When grouping by m2m and more than one record is set, the timeline item appears just on one group.
When grouping by m2m and dragging for changing the time or the group, the changes on the group will not be set, because it could make disappear the records not related with the changes that we want to make.

CarlosRoca13 commented:
I also tried to add the new libraries, but more changes to adjust the code were needed,  so I have chosen to focus on the main problem

chrisandrewmann commented:
@CarlosRoca13 Have given it a quick test and only problem I see is that any M2M field (i.e Assignees on project.task) that have no value set no longer appear on the timeline. Was that intentional as I know there were some restrictions?
Previously and with M2O fields they would appear in category "UNASSIGNED".

chrisandrewmann commented on web_timeline/static/src/js/timeline_renderer.js:
```diff
@@ -0,0 +1,633 @@
+        event_data_transform: function (evt) {
+            const [date_start, date_stop] = this._get_event_dates(evt);
+            let group = evt[this.last_group_bys[0]];
+            if (group && group instanceof Array) {
```
To fix issue where events are not displayed under UNASSIGNED group when using a M2M field.
Add this small change:

`            if (group && group instanceof Array && group.length > 0) {`

LudLaf replied:
Suggested change `if (group && group instanceof Array && group.length > 0) {` added.

Commit:
[IMP] web_timeline: allow UNASSIGNED grouping when using M2M field

alexis-via committed:
[FIX] web_timeline: order group_by field

Fix bug #2266: before this fix, the lines of the timeline view were always sorted by alphabetical order, ignoring the native order of the model on which the group_by field points to.

Fix duplicate entries in the 'fields' argument when the 'colors' parameters uses the same field in multiple conditions.

alexis-via committed:
Update vis-timeline-graph2d lib: 7.3.4 -> 7.7.0

CarlosRoca13 commented:
Added fix and updated libraries as done in PR https://github.com/OCA/web/pull/2268

pedrobaeza commented:
Have you tested? I find weird behaviors when testing it in v14.

CarlosRoca13 commented:
Yes, and I didn't found that problems...

pedrobaeza commented:
/ocabot migration web_timeline
/ocabot merge nobump
